File: javascript/src/test-elasticsearch.js
#!/usr/bin/env node

/**
 * Elasticsearch Verbindungstest für die Schulungsumgebung
 * OS-unabhängig: Windows, Linux, macOS
 * 
 * Dieses Programm testet die Verbindung zu Elasticsearch und zeigt
 * grundlegende Cluster-Informationen an.
 */

import { Client } from '@elastic/elasticsearch';
import dotenv from 'dotenv';
import path from 'path';
import { fileURLToPath } from 'url';
import { existsSync } from 'fs';

// ES Module __dirname equivalent
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

/**
 * Lädt die Umgebungsvariablen aus der .env Datei
 * Sucht in: ../../.env (übergeordnetes Verzeichnis)
 */

function loadEnvironment() {
    const envPath = path.join(__dirname, '..', '..', '.env');

    if (existsSync(envPath)) {
        dotenv.config({ path: envPath });
        return envPath;
    }

    // Keine .env gefunden, verwende Systemumgebungsvariablen
    console.log('ℹ️  Hinweis: Keine .env Datei gefunden, verwende Systemumgebungsvariablen');
    return null;
}

// Lade Umgebungsvariablen
const envPath = loadEnvironment();


/**
 * Erstellt einen Elasticsearch Client mit SSL-Unterstützung
 */
function createElasticsearchClient() {
    const host = process.env.ELASTIC_HOST;
    const user = process.env.ELASTIC_USER;
    const password = process.env.ELASTIC_PASSWORD;
    const verifyCerts = false;

    const client = new Client({
        node: host,
        auth: {
            username: user,
            password: password
        },
        tls: {
            rejectUnauthorized: verifyCerts
        }
    });

    return client;
}

/**
 * Testet die Verbindung zu Elasticsearch und zeigt Cluster-Status
 */
async function testElasticsearchConnection() {
    console.log('='.repeat(60));
    console.log('ELASTICSEARCH VERBINDUNGSTEST');
    console.log(`Schulungsumgebung - JavaScript (${process.platform})`);
    console.log('='.repeat(60));
    
    try {
        // Erstelle Elasticsearch Client
        const client = createElasticsearchClient();
        
        console.log('🔍 Teste Elasticsearch Verbindung...');
        console.log(`Host: ${process.env.ELASTIC_HOST}`);
        console.log(`User: ${process.env.ELASTIC_USER}`);
        console.log('-'.repeat(50));
        
        // Teste Verbindung und hole Cluster-Info
        const info = await client.info();
        console.log('✅ Verbunden mit Elasticsearch');
        console.log(`   Version: ${info.version.number}`);
        console.log(`   Cluster: ${info.cluster_name}`);
        console.log(`   Lucene Version: ${info.version.lucene_version}`);
        
        // Hole Cluster Health
        const health = await client.cluster.health();
        console.log(`🔍 Cluster-Zustand: ${health.status.toUpperCase()}`);
        console.log(`   Anzahl Nodes: ${health.number_of_nodes}`);
        console.log(`   Anzahl Daten-Nodes: ${health.number_of_data_nodes}`);
        
        // Teste Index-Liste
        const indices = await client.cat.indices({ format: 'json' });
        console.log(`📋 Anzahl Indices: ${indices.length}`);
        
        if (indices.length > 0) {
            console.log('   Verfügbare Indices:');
            indices.slice(0, 5).forEach(index => {
                console.log(`     - ${index.index} (${index['docs.count']} Dokumente)`);
            });
            if (indices.length > 5) {
                console.log(`     ... und ${indices.length - 5} weitere`);
            }
        }
        
        console.log('\n🎉 Elasticsearch Verbindungstest erfolgreich!');
        process.exit(0);
        
    } catch (error) {
        console.error('❌ Fehler bei Elasticsearch Verbindung:');
        console.error(`   ${error.message}`);
        console.error('\n💡 Mögliche Lösungen:');
        console.error('   1. Stellen Sie sicher, dass Elasticsearch läuft (docker-compose up)');
        console.error('   2. Prüfen Sie die .env Datei im Projektverzeichnis');
        console.error('   3. Prüfen Sie die Netzwerkverbindung');
        console.error('   4. Warten Sie bis Elasticsearch vollständig gestartet ist');
        
        process.exit(1);
    }
}

// Führe Test aus
testElasticsearchConnection();

File: python/src/test_elasticsearch.py
#!/usr/bin/env python3
"""
Elasticsearch Verbindungstest für die Schulungsumgebung
OS-unabhängig: Windows, Linux, macOS
"""

import os
import sys
from pathlib import Path
from elasticsearch import Elasticsearch
from dotenv import load_dotenv

def load_environment():
    """
    Lädt die Umgebungsvariablen aus der .env Datei
    Sucht in: ../../.env (übergeordnetes Verzeichnis)
    """
    script_dir = Path(__file__).parent
    env_path = script_dir.parent.parent / '.env';
    
    if env_path.exists():
        load_dotenv(dotenv_path=env_path)
        return str(env_path)
    
    # Keine .env gefunden, verwende Systemumgebungsvariablen
    print("ℹ️  Hinweis: Keine .env Datei gefunden, verwende Systemumgebungsvariablen")
    return None

# Lade Umgebungsvariablen
env_file = load_environment()

def test_elasticsearch_connection():
    """Testet die Verbindung zu Elasticsearch und zeigt Cluster-Status"""
    
    try:
        # Erstelle Elasticsearch Client
        es = Elasticsearch(
            os.environ["ELASTIC_HOST"],
            basic_auth=(os.environ["ELASTIC_USER"], os.environ["ELASTIC_PASSWORD"]),
            verify_certs=False,
            ssl_show_warn=False
        )
        
        print("🔍 Teste Elasticsearch Verbindung...")
        print(f"Host: {os.environ['ELASTIC_HOST']}")
        print(f"User: {os.environ['ELASTIC_USER']}")
        print("-" * 50)
        
        # Teste Verbindung und hole Cluster-Info
        info = es.info()
        print(f"✅ Verbunden mit Elasticsearch")
        print(f"   Version: {info['version']['number']}")
        print(f"   Cluster: {info['cluster_name']}")
        print(f"   Lucene Version: {info['version']['lucene_version']}")
        
        # Hole Cluster Health
        health = es.cluster.health()
        print(f"🔍 Cluster-Zustand: {health['status'].upper()}")
        print(f"   Anzahl Nodes: {health['number_of_nodes']}")
        print(f"   Anzahl Daten-Nodes: {health['number_of_data_nodes']}")
        
        # Teste Index-Liste
        indices = es.cat.indices(format='json')
        print(f"📋 Anzahl Indices: {len(indices)}")
        
        if indices:
            print("   Verfügbare Indices:")
            for idx in indices[:5]:  # Zeige nur die ersten 5
                print(f"     - {idx['index']} ({idx['docs.count']} Dokumente)")
            if len(indices) > 5:
                print(f"     ... und {len(indices) - 5} weitere")
        
        print("\n🎉 Elasticsearch Verbindungstest erfolgreich!")
        return True
        
    except Exception as e:
        print(f"❌ Fehler bei Elasticsearch Verbindung:")
        print(f"   {str(e)}")
        print("\n💡 Mögliche Lösungen:")
        print("   1. Stellen Sie sicher, dass Elasticsearch läuft (docker-compose up)")
        print("   2. Prüfen Sie die .env Datei im Projektverzeichnis")
        print("   3. Prüfen Sie die Netzwerkverbindung")
        return False

def main():
    """Hauptfunktion"""
    print("=" * 60)
    print("ELASTICSEARCH VERBINDUNGSTEST")
    print(f"Schulungsumgebung - Python ({sys.platform})")
    print("=" * 60)
    
    success = test_elasticsearch_connection()
    
    if success:
        sys.exit(0)
    else:
        sys.exit(1)

if __name__ == "__main__":
    main()
